SimPEG/utils/model_utils.py

```diff
@@ -195,3 +198,95 @@ def depth_weighting(
         wz = wz[active_cells]
 
     return wz / np.nanmax(wz)
+
+
+def distance_weighting(
+    mesh: discretize.BaseMesh,
+    reference_locs: np.ndarray,
+    active_cells: np.ndarray | None = None,
+    exponent: float = 2.0,
+    threshold: float = None,
+):
+    r"""
+    Construct diagonal elements of a distance weighting matrix
+
+    Builds the model weights following the distance weighting strategy, a method
+    to generate weights based on the distance between mesh cell centers and some
+    reference location(s).
+    Use these weights in regularizations to counteract the natural decay of
+    potential field data with distance.
+
+    Parameters
+    ----------
+    mesh : discretize.base.BaseMesh
+        Discretized model space.
+    reference_locs : float or (n, ndim) numpy.ndarray
+        Reference location for the distance weighting.
+        It can be a ``float``, which value is the component for
+        the reference location.
+        Or it can be a 2d array, with multiple reference locations, where each
+        row should contain the coordinates of a single location point in the
+        following order: _x_, _y_, _z_ (for 3D meshes) or _x_, _z_ (for 2D
+        meshes).
+        The coordinate of the reference location, usually the receiver locations
+    active_cells : (mesh.n_cells) numpy.ndarray of bool, optional
+        Index vector for the active cells on the mesh.
+        If ``None``, every cell will be assumed to be active.
+    exponent : float, optional
+        Exponent parameter for distance weighting.
+        The exponent should match the natural decay power of the potential
+        field. For example, for gravity acceleration, set it to 2; for magnetic
+        fields, to 3.
+    threshold : float or None, optional
+        Threshold parameters used in the distance weighting.
+        If ``None``, it will be set to half of the smallest cell width.
+
+    Returns
+    -------
+    (n_active) numpy.ndarray
+        Normalized distance weights for the mesh at every active cell as
+        a 1d-array.
+    """
+
+    active_cells = (
+        np.ones(mesh.n_cells, dtype=bool) if active_cells is None else active_cells
+    )
+
+    # Default threshold value
+    if threshold is None:
+        threshold = 0.5 * mesh.h_gridded.min()
+
+    reference_locs = np.asarray(reference_locs)
+
+    # Calculate distance from receiver locations
+    # reference_locs is a scalar
+    if reference_locs.ndim < 2:
+        distance = np.abs(mesh.cell_centers[:, -1] - reference_locs)
+
+    else:
+        cell_centers = mesh.cell_centers[active_cells]
+        cell_volumes = mesh.cell_volumes[active_cells]
+        n, d = cell_centers.shape
+        t, d1 = reference_locs.shape
+
+        if not d == d1:
+            raise Exception("vectors must have same number of columns")
+
+        # vectorized distance calculations
+        distance = (
```

Comment:
This one calculation requires several times the sensitivity matrix size:

It creates one when you do:
`np.dot((cell_centers**2, np.ones((d, t))  #(n_cell, 3) * (3, n_rx) = (n_cell, n_rx)`

It creates another when you do:
`np.dot(np.ones([n, d]), (reference_locs**2.0).T) #(n_cell, 3) * (3, n_rx) = (n_cell, n_rx)`

It creates a third when you do:
`np.dot(cell_centers, reference_locs.T)  #(n_cell, 3) * (3, n_rx) = (n_cell, n_rx)`


it creates another when you add a few together another when you do subtract (at which point some of the intermediaries can be garbage collected).

Reply:
what would be your preferred path forward?

Reply:
The only way to get around it is to painfully do a python loop over the receiver locations... I think this is just one of those times where you can't avoid it and must take the time hit (alternatively this would make a great candidate for `numba` acceleration @santisoler).

Reply:
`for` loops across datapoint to keep `n_cells` array?

Reply:
I would agree that a parallel loop would make sense here, since we are just summing all the rows at the end.

Reply:
yeah, usually `n_rx` << `n_cells` so if we have to do a loop, make it do as much work as possible

Reply:
took a stab at it (not familiar with numba coding). See what you think